transformer.cc

```diff
@@ -547,7 +524,210 @@ class V1ToOld final : public TransformerBase {
+  }
+
+  WireFormat From() const { return WireFormat::kV1; }
+  WireFormat To() const { return WireFormat::kOld; }
+
+  zx_status_t TransformUnion(const fidl::FidlCodedUnion& src_coded_union,
+                             const fidl::FidlCodedUnion& dst_coded_union,
+                             const Position& position, TraversalResult* out_traversal_result) {
+    assert(src_coded_union.field_count == dst_coded_union.field_count);
+
+    // Read: extensible-union ordinal.
+    auto src_xunion = src_dst->Read<const fidl_xunion_t>(position);
+    uint32_t xunion_ordinal = src_xunion->tag;
+
+    if (src_xunion->padding != static_cast<decltype(src_xunion->padding)>(0)) {
+      return Fail(ZX_ERR_BAD_STATE, "xunion padding is non-zero");
+    }
+
+    // TODO(apang): Validate src_xunion->envelope->num_bytes
+    // TODO(apang): Validate src_xunion->envelope->num_handles
+
+    switch (src_xunion->envelope.presence) {
+      case FIDL_ALLOC_PRESENT:
+        // OK
+        break;
+      case FIDL_ALLOC_ABSENT:
+        return Fail(ZX_ERR_BAD_STATE, "xunion envelope is invalid FIDL_ALLOC_ABSENT");
+      default:
+        return Fail(ZX_ERR_BAD_STATE, "xunion envelope presence neither FIDL_ALLOC_PRESENT nor FIDL_ALLOC_ABSENT");
+    }
+
+
+    // Retrieve: flexible-union field (or variant).
+    bool src_field_found = false;
+    uint32_t src_field_index = 0;
+    const fidl::FidlUnionField* src_field = nullptr;
+    for (/* src_field_index needed after the loop */;
+         src_field_index < src_coded_union.field_count; src_field_index++) {
+      const fidl::FidlUnionField* candidate_src_field = &src_coded_union.fields[src_field_index];
+      if (candidate_src_field->xunion_ordinal == xunion_ordinal) {
+        src_field_found = true;
+        src_field = candidate_src_field;
+        break;
+      }
+    }
+    if (!src_field_found) {
+      return Fail(ZX_ERR_BAD_STATE, "ordinal has no corresponding variant");
+    }
+
+    const fidl::FidlUnionField& dst_field = dst_coded_union.fields[src_field_index];
+
+    // Write: static-union tag, and pad (if needed).
+    switch (dst_coded_union.data_offset) {
+      case 4:
+        src_dst->Write(position, src_field_index);
+        break;
+      case 8:
+        src_dst->Write(position, static_cast<uint64_t>(src_field_index));
+        break;
+      default:
+        assert(false && "static-union data offset can only be 4 or 8");
+    }
+
+    // Transform: static-union field (or variant).
+    auto field_position = Position{
+        position.src_out_of_line_offset,
+        position.src_out_of_line_offset
+                                + AlignedInlineSize(src_field->type, WireFormat::kOld),
+        position.dst_inline_offset + dst_coded_union.data_offset,
+        position.dst_out_of_line_offset,
+    };
+    uint32_t dst_field_size = dst_coded_union.size - dst_coded_union.data_offset;
+
+    zx_status_t status = Transform(src_field->type, field_position, dst_field_size, out_traversal_result);
```

Comment:
I think this is correct (evidenced by the tests :), since all out-of-line data should be aligned to 8 bytes.

If we want to be extra-safe, we could assert that out_traversal_result->out_of_line_size % 8 == 0 (FIDL_ALIGNed), but I don't think that's necessary.